src/lib.rs

```diff
@@ -258,6 +266,185 @@ pub fn recv_msg(sock: &TsnSocket, msg: &mut msghdr) -> Result<isize, String> {
+        rx_filter: libc::HWTSTAMP_FILTER_NONE as i32,
+        flags: 0,
+    };
+
+    let mut ifr_name: [libc::c_char; libc::IFNAMSIZ] = [0; libc::IFNAMSIZ];
+    for (source, target) in interface_name.as_bytes().iter().zip(ifr_name.iter_mut()) {
+        *target = *source as libc::c_char;
+    }
+
+    let ifreq = libc::ifreq {
+        ifr_name,
+        ifr_ifru: libc::__c_anonymous_ifr_ifru {
+            ifru_data: (&mut ts_cfg as *mut _) as *mut libc::c_char,
+        },
+    };
+
+    let err = unsafe {
+        // Not useless conversion because aarch64 has different type
+        #[allow(clippy::useless_conversion)]
+        libc::ioctl(sockfd, libc::SIOCSHWTSTAMP.try_into().unwrap(), &ifreq)
+    };
+    if err < 0 {
+        // XXX: While ioctl failed, SW timestamp is still enabled.
+        eprintln!("ioctl SIOCSHWTSTAMP failed: {}", Error::last_os_error());
+        eprintln!("But SW timestamp by kernel is still enabled.")
+    }
+
+    Ok(())
+}
+
+pub fn get_tx_timestamp(sock: &TsnSocket) -> Result<time::Timespec, Error> {
+    let sockfd = sock.fd;
+
+    let buf: [u8; 256] = [0u8; 256];
+    let buflen = std::mem::size_of_val(&buf);
+
+    let control: [u8; 256] = [0u8; 256];
+
+    let iov = libc::iovec {
+        iov_base: buf.as_ptr() as *mut libc::c_void,
+        iov_len: buflen,
+    };
+
+    let msg: libc::msghdr = unsafe {
+        // Avoid private field not provided error
+        let mut msg: libc::msghdr = std::mem::MaybeUninit::zeroed().assume_init();
+
+        msg.msg_name = std::ptr::null_mut();
+        msg.msg_namelen = 0;
+        msg.msg_iov = &iov as *const _ as *mut libc::iovec;
+        msg.msg_iovlen = 1;
+        msg.msg_control = control.as_ptr() as *mut libc::c_void;
+        msg.msg_controllen = {
+            // aarch64 has msg_controllen as u32, not usize
+            #[allow(clippy::useless_conversion)]
+            std::mem::size_of_val(&control).try_into().unwrap()
+        };
+        msg.msg_flags = 0;
+        msg
+    };
+
+    let pfd = libc::pollfd {
+        fd: sockfd,
+        events: libc::POLLPRI,
+        revents: 0,
+    };
+
+    let res = unsafe { libc::poll(&pfd as *const _ as *mut libc::pollfd, 1, 1000) };
+
+    match res {
+        0 => {
+            println!("poll timeout");
+            return Err(Error::new(ErrorKind::TimedOut, "poll timeout"));
+        }
+        res if res < 0 => {
+            println!("poll error: {}", Error::last_os_error());
+            return Err(Error::last_os_error());
+        }
+        _ => {}
+    }
+
+    // XXX: IDK why but this doesn't work on NXP
+    // Commenting this out for now
+    // if !(pfd.revents & libc::POLLPRI) != 0 {
+    //     return Err(Error::new(ErrorKind::Other, format!("unexpected revents {}", pfd.revents)));
+    // }
+
+    // Poll done. Now read the timestamp
+
+    let cnt = unsafe {
+        libc::recvmsg(
+            sockfd,
+            &msg as *const _ as *mut libc::msghdr,
```

Comment:
Please, consider the comment for line 380 to be applied here:

356  `let mut msg: ..`
405  `&mut msg,`